server/src/main/java/com/objectcomputing/checkins/services/agenda_item/AgendaItemServicesImpl.java

```diff
@@ -1,82 +1,160 @@
 
-    @Inject
     private CheckInRepository checkinRepo;
-    @Inject
-    private AgendaItemRepository agendaItemRepo;
-    @Inject
+    private AgendaItemRepository agendaItemRepository;
     private MemberProfileRepository memberRepo;
+    private SecurityService securityService;
+    private CurrentUserServices currentUserServices;
+
+    public AgendaItemServicesImpl(CheckInRepository checkinRepo, AgendaItemRepository agendaItemRepository,
+                                   MemberProfileRepository memberRepo, SecurityService securityService,
+                                   CurrentUserServices currentUserServices) {
+        this.checkinRepo = checkinRepo;
+        this.agendaItemRepository = agendaItemRepository;
+        this.memberRepo = memberRepo;
+        this.securityService = securityService;
+        this.currentUserServices = currentUserServices;
+    }
+
 
+    @Override
     public AgendaItem save(AgendaItem agendaItem) {
         AgendaItem agendaItemRet = null;
+        String workEmail = securityService != null ? securityService.getAuthentication().get().getAttributes().get("email").toString() : null;
+        MemberProfile currentUser = workEmail != null ? currentUserServices.findOrSaveUser(null, workEmail) : null;
+        Boolean isAdmin = securityService != null ? securityService.hasRole(RoleType.Constants.ADMIN_ROLE) : false;
+
         if (agendaItem != null) {
             final UUID checkinId = agendaItem.getCheckinid();
             final UUID createById = agendaItem.getCreatedbyid();
-            if (checkinId == null || createById == null) {
-                throw new AgendaItemBadArgException(String.format("Invalid agendaItem %s", agendaItem));
-            } else if (agendaItem.getId() != null) {
-                throw new AgendaItemBadArgException(String.format("Found unexpected id %s for agenda item", agendaItem.getId()));
-            } else if (!checkinRepo.findById(checkinId).isPresent()) {
-                throw new AgendaItemBadArgException(String.format("CheckIn %s doesn't exist", checkinId));
-            } else if (!memberRepo.findById(createById).isPresent()) {
-                throw new AgendaItemBadArgException(String.format("Member %s doesn't exist", createById));
+            CheckIn checkinRecord = checkinRepo.findById(checkinId).orElse(null);
+            Boolean isCompleted = checkinRecord != null ? checkinRecord.isCompleted() : null;
+            final UUID pdlId = checkinRecord != null ? checkinRecord.getPdlId() : null;
+            validate(checkinId == null || createById == null, "Invalid agenda item %s", agendaItem);
+            validate(agendaItem.getId() != null, "Found unexpected id %s for agenda item", agendaItem.getId());
+            validate(checkinRepo.findById(checkinId).isEmpty(), "CheckIn %s doesn't exist", checkinId);
+            validate(memberRepo.findById(createById).isEmpty(), "Member %s doesn't exist", createById);
+            if (!isAdmin && isCompleted) {
```

Comment:
If user is not admin and if checkin is complete, save operation is not allowed. Shouldn't we throw exception directly for this?

Reply:
updated to throw exception directly.